### PR TITLE
[SYCL-MLIR] Lower to pointers with correct address spaces

### PR DESCRIPTION
**Pull request by victor-eds**
Before this change, we were assuming some address spaces when lowering `sycl` operations to `llvm`.

Now, when getting a pointer to a member of a struct pointed by a pointer in a given address space, we return a pointer in such address space unless the optional `targetAddressSpace` argument is passed. In that case, we cast to the given address space.

When offsetting an accessor, we obtain the address space from a mapping from target mode to address space, as defined [in the SYCL standard](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:opencl:native-interop-kernel). Note that DPC++ representation of the `device`/`global` and `constant`  (deprecated) address spaces is the same (1).

### CODE REVIEW

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 1]
```diff
@@ -1069,12 +1084,23 @@ class AccessorSubscriptPattern
...
+                                           .getTargetMode());
+    const auto gepPtrTy =
+        LLVM::LLVMPointerType::get(ptrTy.getElementType(), addressSpace);
+    const auto ptr = GetMemberPattern<AccessorGetPtr>::loadValue(builder, loc,
+                                                                 gepPtrTy, acc);
+    const Value gep = builder.create<LLVM::GEPOp>(loc, gepPtrTy, ptr, index,
+                                                  /*inbounds*/ true);
+    return ptrTy == gepPtrTy
+               ? gep
+               : builder.create<LLVM::AddrSpaceCastOp>(loc, ptrTy, gep);
```
Comment (sommerlukas) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:1103]: Same comment as above about order of `gep` and AS cast.

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-flat-to-llvm.mlir [hunk 2]
```diff
@@ -0,0 +1,163 @@
+// RUN: sycl-mlir-opt -split-input-file -convert-sycl-to-llvm %s | FileCheck %s
+
+//===-------------------------------------------------------------------------------------------------===//
```
Comment (sommerlukas) [inline mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-flat-to-llvm.mlir:3]: NIT: Exceeds column width. 

Applies to most of these lines in this file.

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-flat-to-llvm.mlir [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+// CHECK-NEXT:    }
+func.func @test(%id: memref<?x!sycl_id_3_>, %idx: i32) -> memref<?xi64, 4> {
+  %0 = sycl.id.get %id[%idx] { ArgumentTypes = [memref<?x!sycl_id_3_>, i32], FunctionName = @"operator[]", MangledFunctionName = @"operator[]", TypeName = @"id" }  : (memref<?x!sycl_id_3_>, i32) -> memref<?xi64, 4>
+  return %0 : memref<?xi64, 4>
+}
+
+// -----
+
+//===-------------------------------------------------------------------------------------------------===//
+// sycl.accessor.subscript with scalar offset and 1D accessor//===-------------------------------------------------------------------------------------------------===//
```
Comment (sommerlukas) [inline mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-flat-to-llvm.mlir:44]: NIT: Formatting.

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 4]
```diff
@@ -244,12 +259,20 @@ class GetMemberPatternBase {
...
-  /// Returns a value of the type pointed by \p ty to a member of the struct
-  /// pointed by \p ptr.
+    const auto addressSpace =
+        ptr.getType().cast<LLVM::LLVMPointerType>().getAddressSpace();
+    const auto ptrTy = LLVM::LLVMPointerType::get(ty, addressSpace);
+    const Value gep = builder.create<LLVM::GEPOp>(loc, ptrTy, ptr, indices,
+                                                  /*inbounds*/ true);
+    return !targetAddressSpace || *targetAddressSpace == addressSpace
+               ? gep
+               : builder.create<LLVM::AddrSpaceCastOp>(
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp]: Mimmicing clang: we will keep this order for casting output of `accessor::operator[]` and apply the one you suggest when accessing members of a given struct.

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-flat-to-llvm.mlir (hunk 2 above)
Comment (victor-eds) [reply mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-flat-to-llvm.mlir]: We don't have a strict column width limit for lit tests. I'd argue it's more readable this way. Do you have issues visualizing this in your Emacs?

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-flat-to-llvm.mlir (hunk 3 above)
Comment (victor-eds) [reply mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-flat-to-llvm.mlir]: Ups, fixed!

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-flat-to-llvm.mlir (hunk 2 above)
Comment (sommerlukas) [reply mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-flat-to-llvm.mlir]: It wraps in my editor, which looks a bit weird. LLVM coding guidelines also suggest column width of 80. I get that limiting the column width of `CHECK` lines does not make much sense, but for these comment lines, we could stick with the limit, at least the newly added ones.

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 1 above)
Comment (sommerlukas) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:1110]: Here, having `gep` before the AS cast matches clang's behavior, therefore this does not need be changed.

**Review by sommerlukas — APPROVED** [review]

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 1 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:1110]: **Offline conversation**: Changed comment above. Will keep this as is to mimic what `clang` does.

**Review by etiotto — APPROVED** [review]

**Comment by whitneywhtsang** [comment]
Do you know if 
```
Failed Tests (2):
  SYCL :: InvokeSimd/Spec/ImplicitSubgroup/simd_mask.cpp
  SYCL :: InvokeSimd/Spec/simd_mask.cpp
```
are regression?

**Review by Naghasan — APPROVED** [review]

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**